buildSrc/src/main/kotlin/codegen/math/GenerateMath.kt

```diff
@@ -0,0 +1,1027 @@
+							.zipWithNext()
+							.joinToString(",\n\t", "\n\t", "\n") { (c1, c2) ->
+								"$c1 * other.$c2 - other.$c1 * $c2"
+							}
+						statement("return %T($args)", type)
+					}
+				}
+			}.writeTo(commonDir.get().asFile)
+		}
+	}
+
+	private fun vectorTypes() {
+		vectorTypes.forEach { (type, baseType, componentCount) ->
+			val mutableVectorType = ClassName(packageName, "Mutable${type.simpleName}")
+			val componentNames = allComponentNames.take(componentCount!!)
+			baseType ?: error("null base type invalid for vectors")
+
+			val (zero) = literal("0", baseType, null)
+
+			buildFile(packageName, type.simpleName) {
+				indent("\t")
+
+				import("kotlin.math", "sqrt")
+
+				buildClass(type) {
+					modifiers(SEALED)
+
+					companionObject()
+
+					componentNames.forEach {
+						property(it, baseType, ABSTRACT)
+					}
+
+					function("get") {
+						modifiers(OPERATOR)
+						parameter("index", INT)
+						controlFlow("when (index)") {
+							componentNames.forEachIndexed { i, it ->
+								statement("$i -> $it")
+							}
+							statement("else -> throw IndexOutOfBoundsException()")
+						}
+					}
+
+					if (baseType != BOOLEAN) {
+						property("length", baseType) {
+							getter {
+								val code = when (baseType) {
+									INT, LONG ->
+										"return sqrt((this dot this).toDouble()).to${baseType.simpleName}()"
+									U_INT, U_LONG ->
+										"return sqrt((this dot this).toDouble()).to${baseType.simpleName.drop(1)}().to${baseType.simpleName}()"
+									else ->
+										"return sqrt(this dot this)"
+								}
+								statement(code)
+							}
+						}
+
+						property("lengthSquared", baseType) {
+							getter {
+								statement("return this dot this")
+							}
+						}
+
+						function("normalized") {
+							returns(type)
+							statement("val length = length")
+							val args = componentNames.joinToString { "$it / length" }
+							statement("return %T($args)", type)
+						}
+
+						for ((name, op) in arithmetics) {
+							function(name) {
+								modifiers(OPERATOR)
+								parameter("scalar", baseType)
+								returns(type)
+								val args = componentNames.joinToString(",\n\t", "\n\t", "\n") { "$it $op scalar" }
+								statement("return %T($args)", type)
+							}
+
+							function(name) {
+								modifiers(OPERATOR)
+								parameter("other", type)
+								returns(type)
+								val args = componentNames.joinToString(",\n\t", "\n\t", "\n") { "$it $op other.$it" }
+								statement("return %T($args)", type)
+							}
+						}
+					}
+				}
+
+				buildClass(mutableVectorType) {
+					superclass(type)
+
+					companionObject()
+
+					primaryConstructor {
+						componentNames.forEach { name ->
+							parameter(name, baseType)
+						}
+						callSuperConstructor()
+					}
+
+					secondaryConstructor {
+						parameter("scalar", baseType)
+						callThisConstructor(*componentNames.map { "scalar" }.toTypedArray())
+					}
+
+					secondaryConstructor {
```

Comment:
This secondary ctor can be merged into the previous one. By giving `scalar` a default value of 0.0f.

Reply:
This actually breaks in the compiler by creating a weird redeclaration of the ctor (at least it did when I tried it before). I'll check again.

Reply:
yeah throws platform declaration clash on jvm compile. Seems to be related to the unsigned  types (since they're inlined as their signed counterparts)

Reply:
Oh that's unfortunate. Might be worth filing an issue on YouTrack or something.

Reply:
Definitely. I'll need to create a minimum working (or failing as it were) example first, tho

Reply:
https://youtrack.jetbrains.com/issue/KT-37405